File: cat-core/src/main/java/com/dianping/cat/message/internal/MessageId.java
package com.dianping.cat.message.internal;

import java.util.List;

import com.site.helper.Splitters;

public class MessageId {
	private static final long VERSION1_THRESHOLD = 1325347200000L; // Jan. 1, 2012

	private String m_domain;

	private String m_ipAddressInHex;

	private long m_timestamp;

	private int m_index;

	MessageId(String domain, String ipAddressInHex, long timestamp, int index) {
		m_domain = domain;
		m_ipAddressInHex = ipAddressInHex;
		m_timestamp = timestamp;
		m_index = index;
	}

	public static MessageId parse(String messageId) {
		List<String> list = Splitters.by('-').split(messageId);
		int len = list.size();

		if (len >= 4) {
			String ipAddressInHex = list.get(len - 3);
			long timestamp = Long.parseLong(list.get(len - 2));
			int index = Integer.parseInt(list.get(len - 1));
			String domain;

			if (len > 4) { // allow domain contains '-'
				StringBuilder sb = new StringBuilder();

				for (int i = 0; i < len - 3; i++) {
					if (i > 0) {
						sb.append('-');
					}

					sb.append(list.get(i));
				}

				domain = sb.toString();
			} else {
				domain = list.get(0);
			}

			return new MessageId(domain, ipAddressInHex, timestamp, index);
		}

		throw new RuntimeException("Invalid message id format: " + messageId);
	}

	public String getDomain() {
		return m_domain;
	}

	public int getIndex() {
		return m_index;
	}

	public String getIpAddressInHex() {
		return m_ipAddressInHex;
	}

	public long getTimestamp() {
		if (m_timestamp > VERSION1_THRESHOLD) {
			return m_timestamp;
		} else {
			return m_timestamp * 3600 * 1000L;
		}
	}

	public int getVersion() {
		if (m_timestamp > VERSION1_THRESHOLD) {
			return 1;
		} else {
			return 2;
		}
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder(m_domain.length() + 32);

		sb.append(m_domain);
		sb.append('-');
		sb.append(m_ipAddressInHex);
		sb.append('-');
		sb.append(m_timestamp);
		sb.append('-');
		sb.append(m_index);

		return sb.toString();
	}
}

File: lib/csharp/src/CatClient/Message/Internals/MessageId.cs
using System.Globalization;
using System;
using System.Collections.Generic;
using System.Text;

namespace Org.Unidal.Cat.Message.Internals
{
    [Obsolete]
    public class MessageId
    {
        private readonly String _mDomain;

        private readonly int _mIndex;

        private readonly String _mIpAddressInHex;

        private readonly long _mTimestamp;

        internal MessageId(String domain, String ipAddressInHex, long timestamp, int index)
        {
            _mDomain = domain;
            _mIpAddressInHex = ipAddressInHex;
            _mTimestamp = timestamp;
            _mIndex = index;
        }

        public String Domain
        {
            get { return _mDomain; }
        }

        public int Index
        {
            get { return _mIndex; }
        }

        public String IpAddressInHex
        {
            get { return _mIpAddressInHex; }
        }

        public long Timestamp
        {
            get { return _mTimestamp; }
        }

        private static MessageId Parse(String messageId)
        {
            IList<String> list = messageId.Split('-');
            int len = list.Count;

            if (len >= 4)
            {
                String ipAddressInHex = list[len - 3];
                long timestamp = (Int64.Parse(list[len - 2], NumberStyles.Integer));
                int index = Int32.Parse(list[len - 1]);
                String domain;

                if (len > 4)
                {
                    // allow domain contains '-'
                    StringBuilder sb = new StringBuilder();

                    for (int i = 0; i < len - 3; i++)
                    {
                        if (i > 0)
                        {
                            sb.Append('-');
                        }

                        sb.Append(list[i]);
                    }

                    domain = sb.ToString();
                }
                else
                {
                    domain = list[0];
                }

                return new MessageId(domain, ipAddressInHex, timestamp, index);
            }

            throw new Exception("Invalid message id format: " + messageId);
        }

        public override String ToString()
        {
            try
            {
                StringBuilder sb = new StringBuilder(_mDomain.Length + 32);

                sb.Append(_mDomain);
                sb.Append('-');
                sb.Append(_mIpAddressInHex);
                sb.Append('-');
                sb.Append(_mTimestamp);
                sb.Append('-');
                sb.Append(_mIndex);

                return sb.ToString();
            }
            catch (Exception ex)
            {
                Cat.lastException = ex;
                return "";
            }
        }
    }
}